### PR TITLE
Chat: Add a method for iterating commands

### PR DESCRIPTION
**Pull request by mia-pi-git**
Cleans up some code and can be used elsewhere.

### CODE REVIEW

**Comment by Zarel** [comment]
It might be easier to return an array of every handler (which already contains `handler.fullCmd`), which you can then `filter` and `map` to get what you're looking for.

**Comment by mia-pi-git** [comment]
@Zarel thoughts on this API?

File: server/chat-commands/admin.ts [hunk 1]
```diff
@@ -367,11 +352,14 @@ export const commands: ChatCommands = {
 				}
 				this.sendReply("Hotpatching chat commands...");
 
-				const disabledCommands = getDisabledCommands();
+				const disabledCommands: string[] = [];
+				for (const handler of Chat.allCommands()) {
+					if (handler.disabled) disabledCommands.push(`/${handler.fullCmd}`);
+				}
```
Comment (Zarel) [inline server/chat-commands/admin.ts:358]:
```suggestion
				const disabledCommands = Chat.allCommands().filter(c => c.disabled).map(c => `/${c.fullCmd}`);
```

File: server/chat-commands/admin.ts [hunk 2]
```diff
@@ -678,11 +666,14 @@ export const commands: ChatCommands = {
 	lockdown(target, room, user) {
 		this.checkCan('lockdown');
 
-		const disabledCommands = getDisabledCommands();
+		const disabledCommands: string[] = [];
+		for (const handler of Chat.allCommands()) {
+			if (handler?.disabled) disabledCommands.push(`/${handler.fullCmd}`);
+		}
```
Comment (Zarel) [inline server/chat-commands/admin.ts:672]:
```suggestion
		const disabledCommands = Chat.allCommands().filter(c => c.disabled).map(c => `/${c.fullCmd}`);
```

File: server/chat.ts [hunk 3]
```diff
@@ -1627,13 +1628,21 @@ export const Chat = new class {
...
 			}
-			if (typeof entry !== 'function') continue;
+			if (typeof entry === 'string') {
+				const base = commandTable[entry];
+				if (!base) continue;
+				if (!base.aliases) base.aliases = [];
+				base.aliases.push(cmd);
+				continue;
+			}
+ 			if (typeof entry !== 'function') continue;
```
Comment (Zarel) [inline server/chat.ts:1638]:
```suggestion
 		if (typeof entry !== 'function') continue;
```

File: server/chat.ts [hunk 4]
```diff
@@ -1830,6 +1839,19 @@ export const Chat = new class {
...
+			const handler = table[cmd];
+			if (Array.isArray(handler) || !handler || typeof handler === 'string') continue;
+			if (typeof handler === 'object') {
+				results.push(...this.allCommands(handler, `${namespace}${cmd} `));
+				continue;
+			}
+			results.push(handler as AnnotatedChatHandler);
+		}
+		return results;
+	}
```
Comment (Zarel) [inline server/chat.ts:1856]:
```suggestion
	allCommands(table: ChatCommands = Chat.commands) {
		const results: AnnotatedChatHandler[] = [];
		for (const cmd in table) {
			const handler = table[cmd];
			if (Array.isArray(handler) || !handler || typeof handler === 'string') continue;
			if (typeof handler === 'object') {
				results.push(...this.allCommands(handler));
				continue;
			}
			results.push(handler as AnnotatedChatHandler);
		}
		return results;
	}
```

`namespace` appears unused.

File: server/user-groups.ts [hunk 5]
```diff
@@ -159,21 +159,8 @@ export abstract class Auth extends Map<ID, GroupSymbol | ''> {
...
-					const subEntry = (entry as Chat.AnnotatedChatCommands)[subCommand];
-					if (typeof subEntry !== 'function') continue;
-					if (subEntry.hasRoomPermissions) permissions.push(`/${cmd} ${subCommand}`);
-				}
-				continue;
-			}
+		for (const handler of Chat.allCommands()) {
+			if (handler.hasRoomPermissions) permissions.push(`/${handler.fullCmd}`);
 		}
 		return permissions;
```
Comment (Zarel) [inline server/user-groups.ts]:
```ts
return [
  ...ROOM_PERMISSIONS,
  ...Chat.callCommands.filter(c => c.hasRoomPermissions).map(c => c.fullCmd),
];
```

File: server/user-groups.ts [hunk 6]
```diff
@@ -159,23 +159,9 @@ export abstract class Auth extends Map<ID, GroupSymbol | ''> {
 	}
 	static supportedRoomPermissions(room: Room | null = null) {
 		const permissions: string[] = ROOM_PERMISSIONS.slice();
```
Comment (Zarel) [inline server/user-groups.ts:161]:
```suggestion
```

File: server/user-groups.ts [hunk 7]
```diff
@@ -159,23 +159,9 @@ export abstract class Auth extends Map<ID, GroupSymbol | ''> {
...
-					if (!entry.hasRoomPermissions) continue;
-					if (room && typeof entry.requiresRoom === 'string') {
-						if (room.roomid !== entry.requiresRoom) continue;
-					}
-					permissions.push(`/${namespace}${cmd}`);
-				}
-			}
+		for (const handler of Chat.allCommands()) {
+			if (handler.hasRoomPermissions) permissions.push(`/${handler.fullCmd}`);
 		}
```
Comment (Zarel) [inline server/user-groups.ts:164]:
```suggestion
```

File: server/user-groups.ts [hunk 8]
```diff
@@ -159,23 +159,9 @@ export abstract class Auth extends Map<ID, GroupSymbol | ''> {
...
-		checkCmdTable(Chat.commands);
 		return permissions;
```
Comment (Zarel) [inline server/user-groups.ts:165]:
```suggestion
		return [
				...ROOM_PERMISSIONS,
				...Chat.allCommands().filter(c => c.hasRoomPermissions).map(c => `/${c.fullCmd}`),
		];
```

File: server/chat.ts [hunk 9]
```diff
@@ -1629,13 +1630,21 @@ export const Chat = new class {
...
 			}
-			if (typeof entry !== 'function') continue;
+			if (typeof entry === 'string') {
+				const base = commandTable[entry];
+				if (!base) continue;
+				if (!base.aliases) base.aliases = [];
+				base.aliases.push(cmd);
+				continue;
+			}
+ 		if (typeof entry !== 'function') continue;
```
Comment (Zarel) [inline server/chat.ts:1640]:
```suggestion
			if (typeof entry !== 'function') continue;
```

File: server/chat.ts [hunk 10]
```diff
@@ -50,6 +50,7 @@ export type AnnotatedChatHandler = ChatHandler & {
 	fullCmd: string,
 	isPrivate: boolean,
 	disabled: boolean,
+	aliases: string[],
```
Comment (Zarel) [inline server/chat.ts:53]: Should we consider this `string[] | null` instead?

File: server/chat.ts [hunk 11]
```diff
@@ -1629,13 +1630,21 @@ export const Chat = new class {
 			if (typeof entry === 'object') {
 				this.annotateCommands(entry, `${namespace}${cmd} `);
 			}
+			if (typeof entry === 'string') {
+				const base = commandTable[entry];
+				if (!base) continue;
+				if (!base.aliases) base.aliases = [];
+				base.aliases.push(cmd);
```
Comment (Zarel) [inline server/chat.ts:1637]:
```ts
				(base.aliases ||= []).push(cmd);
```

you wish :p

File: server/user-groups.ts [hunk 12]
```diff
@@ -158,25 +158,10 @@ export abstract class Auth extends Map<ID, GroupSymbol | ''> {
...
-					}
-					permissions.push(`/${namespace}${cmd}`);
-				}
-			}
-		}
-		checkCmdTable(Chat.commands);
-		return permissions;
+		return [
+				...ROOM_PERMISSIONS,
+				...Chat.allCommands().filter(c => c.hasRoomPermissions).map(c => `/${c.fullCmd}`),
```
Comment (Zarel) [inline server/user-groups.ts:163]:
```suggestion
			...ROOM_PERMISSIONS,
			...Chat.allCommands().filter(c => c.hasRoomPermissions).map(c => `/${c.fullCmd}`),
```

**Comment by Zarel** [comment]
Thanks!

**Pull request merged**